CLI: Move SSHCommandResult processing to Base class

abalakh opened the pull request:
Core part of #2776

* Added new exception - `CLIReturnCodeError` which contains return_code, stderr and msg (message).
* Created `Base._handle_response()` func which checks `return_code` and raises `CLIReturnCodeError` if needed. Returns `stdout`.
* Injected `Base._handle_response()` in `Base.execute()` (that's the easiest place to inject, no existing commands won't be affected and nothing should be rewritten)
* By default, all CLI commands running with `Base.execute()` will be wrapped by `Base._handle_response()`, if raw response is needed, `Base.execute()` now accepts optional parameter `return_raw_response`.
* Removed `Base._remove_task_status()` in favour of `Base._handle_response()` `ignore_stderr` argument.

abalakh commented on robottelo/cli/base.py:
```diff
@@ -86,6 +86,33 @@ class Base(object):
+                )
+            )
+        if len(response.stderr) != 0 and not ignore_stderr:
+            cls.logger.warning(
```
This should be discussed. I don't know whole backstory, but AFAIK, @elyezer once tried to count on `stderr` and fail tests when `stderr` is not empty. Then it appeared that some commands return some garbage in `stderr` and `remove_task_status()` was created for such commands. But it looks like in the end this approach was totally rejected as now we don't check for `stderr` in the latest test. Is that so?
Depending on the answer, I can see 3 approaches we can use for `stderr`:

1) (Currently applied) Leave `remove_task_status()` for commands as it was for compatibility, in case of `len(stderr) != 0` just throw warning in logs.
2) Remove `remove_task_status()` completely, do not check `stderr` at all
3) Count on `stderr`. Commands which produce some garbage in `stderr` should use `remove_task_status` flag, if flag is not used and `stderr` is not empty - throw an exception. The problem here is that we haven't covered all the commands with checks for `len(stderr)`, so theoretically we can face a lot of new failures. On the other hand, literally all the CLI tests should be refactored because of this PR, so we'll be able to catch all the failures and fix them.

JacobCallahan replied:
I am in favor of keeping the stderr in place. It can provide some useful information in the event of a failure. With that said, we should definitely continue to transition away from checking stderr as a test condition. Basically just use it in the vent of a failure, not as a reason to fail.

abalakh commented on robottelo/cli/factory.py:
```diff
@@ -83,10 +84,10 @@ def create_object(cli_object, options, values):
-    if result.return_code != 0:
+    try:
+        result = cli_object.create(options)
+    except CLICommandExecutionError:
```
This can be discussed, too. I left `CLIFactoryError` here for compatibility reasons, but it can be dropped easily as all the tests will be refactored and `CLIFactoryError` can be replaced with `CLICommandExecutionError`. Should I leave it as it is or drop?

Ichimonji10 replied:
I don't know too much about the details of the CLI test suite, but in general I'd prefer small changes over large changes.

abalakh commented:
`[DO NOT MERGE]` tag as there're some questions. This PR is for `issue_2776` branch, not `master`, so it can be easily merged.
Would be happy to hear some feedback please, especially @elyezer and @Ichimonji10 :smile:

abalakh commented:
One example of what it gives us is #2779

JacobCallahan commented on robottelo/cli/base.py:
```diff
@@ -112,8 +138,8 @@ def create(cls, options=None):
         # Extract new object ID if it was successfully created
-        if len(result.stdout) > 0 and 'id' in result.stdout[0]:
-            obj_id = result.stdout[0]['id']
+        if len(result) > 0 and 'id' in result[0]:
```
it's been a bit since i've looked hard into our code but are we preferring ```if len(something) > 0``` over ```if something```? If so why? Thanks

abalakh replied:
It was originally `if len(something) > 0` so I left it with the same style.
Regarding your question - I guess it depends on type of "`something`". If "`something`" was, for example, string or dictionary - then i guess there's no point in verifying it's length. But `result` here is a list. And for `list` object, imo, it adds readability. For me `if len(list) > 0` == 'if list is not empty and has entries' and it's a little bit clearer than `if list`.
Anyway, it's definitely not an issue and i can update it if needed.

JacobCallahan replied:
no update needed. i've just always preferred the version i mentioned. something we can mull around though.

Ichimonji10 commented on robottelo/cli/base.py:
```diff
@@ -1,28 +1,27 @@
+        self.msg = msg
+
+    def __str__(self):
+        return self.msg
```
Why bother defining `__str__` or `__init__`? Unless there's some specific need, just inherit directly from `Exception`, like so:

```python
class CLICommandExecutionError(Exception):
    """Indicates that a CLI command returned an error after execution."""
```

abalakh replied:
And there's such need - we don't just need some new exception, we need an exception, which will contain `return_code` and `stderr` in it, as we have some tests, which are expecting some specific `return_code` value.
E.g., defining `self.return_code` in `__init__` section will allow us to check for a custom `return_code` like this:
```python
with self.assertRaises(CLICommandExecutionError) as e:
    ContentView.info({'id': 123})
self.assertEqual(e.exception.error_code, 1)
```

Ichimonji10 replied:
Clever.

Ichimonji10 commented on robottelo/cli/base.py:
```diff
@@ -1,28 +1,27 @@
 
 
+class CLICommandExecutionError(Exception):
+    """Indicates that a CLI command returned an error after execution."""
```
Also, can you make the docstring a little more precise? What's the difference between a command returning an error, versus one returning an error "after execution"? What constitutes an error? Just a status code, or does the presence of contents in stderr also warrant raising this exception? I'd be very happy to see a clear definition of when and why this exception is raised.

Ichimonji10 commented on robottelo/cli/base.py:
```diff
@@ -86,6 +85,33 @@ class Base(object):
+    def _handle_response(cls, response, ignore_stderr=None):
+        """Verifies ``return_code`` and ``stderr`` in ``SSHCommandResult``
+        response and throws ``CLICommandExecutionError`` exception if
+        ``return_code`` is not equal to 0.
```
Please re-write this docstring so it has a summary line, body, and directives:

```python
    """Verify foo.

    Check for a non-zero return code or any stderr contents.

    :raises robottelo.cli.base.CLICommandExecutionError: If ...

    """
```

Ichimonji10 commented on robottelo/cli/base.py:
```diff
@@ -86,6 +85,33 @@ class Base(object):
 
     @classmethod
+    def _handle_response(cls, response, ignore_stderr=None):
+        """Verifies ``return_code`` and ``stderr`` in ``SSHCommandResult``
```
Please re-write this docstring so it has a summary line, body, and directives:

```python
"""Verify foo.

Check for a non-zero return code or any stderr contents.

:raises robottelo.cli.base.CLICommandExecutionError: If ...

"""
```

Ichimonji10 commented on robottelo/cli/base.py:
```diff
@@ -86,6 +85,33 @@ class Base(object):
+            raise CLICommandExecutionError(
+                response.return_code,
+                response.stderr,
+                'Command \'{0} {1}\' finished with return_code {2}\n'
```
You can use single and double quotes together to get rid of these back-slashes.

    "Command '{0} {1}' finished with..."

Ichimonji10 commented:
Is there any way to run a CLI command and not have `_handle_response` parse the response? I imagine there'll be cases where it's useful to get at the raw response.

Ichimonji10 commented on robottelo/cli/factory.py:
```diff
@@ -1702,40 +1704,43 @@ def setup_org_for_a_custom_repo(options=None):
+    try:
+        ContentView.publish({u'id': cv_id})
+    except CLICommandExecutionError:
         raise CLIFactoryError('Failed to publish new version of content view')
```
How about including the exception in this error message?

```python
except CLICommandExecutionError as err:
    raise ...  # include err in new exception
```

Ichimonji10 commented on robottelo/cli/factory.py:
```diff
@@ -1702,40 +1704,43 @@ def setup_org_for_a_custom_repo(options=None):
+            u'repository-id': custom_repo['id'],
+        })
+    except CLICommandExecutionError:
         raise CLIFactoryError('Failed to add repository to content view')
```
How about including the exception in this error message?

```python
except CLICommandExecutionError as err:
    raise ...  # include err in new exception
```

Ichimonji10 commented on robottelo/cli/factory.py:
```diff
@@ -1702,40 +1704,43 @@ def setup_org_for_a_custom_repo(options=None):
+    try:
+        Repository.synchronize({'id': custom_repo['id']})
+    except CLICommandExecutionError:
         raise CLIFactoryError('Failed to synchronize repository')
```
How about including the exception in this error message?

```python
except CLICommandExecutionError as err:
    raise ...  # include err in new exception
```

Ichimonji10 commented on robottelo/cli/factory.py:
```diff
@@ -1638,22 +1639,23 @@ def activationkey_add_subscription_to_repo(options=None):
+                    u'subscription-id': subscription['id'],
+                    u'quantity': 1,
+                })
+            except CLICommandExecutionError:
```
How about including this exception in the inner exception?

How about including the exception in this error message?

```python
except CLICommandExecutionError as err:
    raise ...  # include err in new exception
```

Ichimonji10 commented on robottelo/cli/factory.py:
```diff
@@ -1702,40 +1704,43 @@ def setup_org_for_a_custom_repo(options=None):
+            u'organization-id': org_id,
+            u'to-lifecycle-environment-id': env_id,
+        })
+    except CLICommandExecutionError:
```
How about including the exception in this error message?

```python
except CLICommandExecutionError as err:
    raise ...  # include err in new exception
```

Same question applies to rest of module. I'll refrain from commenting more.

JacobCallahan replied:
:)

abalakh commented:
* As we decided not to rely on `stderr`, the only case when `CLICommandExecutionError` can occur is non-zero return code, so i've renamed it to `CLIReturnCodeError`.
* Updated docstrings according to comments.
* Added `CLIReturnCodeError` message to all the `CLIFactoryError` messages.
* Added ability to run CLI command with `Base.execute()` without parsing response by optional `return_raw_response` parameter. Already found it useful in import_ module, as it greatly reduced amount of needed changes in code there.

I think PR is ready for the second cycle of review :smile:

Ichimonji10 commented:
Please update the commit message.

Ichimonji10 commented on robottelo/cli/base.py:
```diff
@@ -86,6 +92,40 @@ class Base(object):
+         ``stderr`` is not empty.
+        :returns: contents of ``stdout``.
+        :raises robottelo.cli.base.CLIReturnCodeError: If return code is
+         different from zero.
```
Can Sphinx parse one-space indentation levels? I'm not sure. Regardless, can you please update to:

```
        :param response: a ``SSHCommandResult`` object, returned by
            :mod:`robottelo.ssh.command`.
        :param ignore_stderr: indicates whether to throw a warning in logs if
            ``stderr`` is not empty.
        :returns: contents of ``stdout``.
        :raises robottelo.cli.base.CLIReturnCodeError: If return code is
            different from zero.
```

Nitpick.

Ichimonji10 commented on robottelo/cli/factory.py:
```diff
@@ -83,23 +84,24 @@ def create_object(cli_object, options, values):
                 cli_object.__name__,
-                json.dumps(options, indent=2, sort_keys=True)
+                json.dumps(options, indent=2, sort_keys=True),
+                err.msg,
```
Thanks for adding this error message to the raised exception. I'm sure it'll aid in debugging when failures happen.

Ichimonji10 commented:
ACK. Fix the nitpick if you can, but it's not a blocker. More importantly, do we have anything to show that the updated code works correctly?

abalakh commented:
@Ichimonji10 Done, thanks for review. As an example there is #2779. I've re-run tests after the latest changes:
```
nosetests tests/foreman/cli/test_partitiontable.py
......
----------------------------------------------------------------------
Ran 6 tests in 93.045s

OK
```
Also, locally i've already prepared about 6-8 updated tests files, just waiting for this PR to be merged and will start spamming with updated tests PRs :smile:

Ichimonji10 commented:
Great! Thank you. :smile:

elyezer commented on robottelo/cli/base.py:
```diff
@@ -1,28 +1,34 @@
+    def __init__(self, return_code, stderr, msg):
+        self.return_code = return_code
+        self.stderr = stderr
+        self.msg = msg
```
Should not call the super __init__?

elyezer commented:
ACK
